pkg/relayer/session/persistence.go

```diff
@@ -0,0 +1,261 @@
+			return nil
+		}
+
+		logger.Error().Err(err).Msg("failed to delete outdated session tree store")
+		return err
+	}
+
+	// Delete the persisted session tree metadata
+	if err := rs.sessionsMetadataStore.Delete([]byte(sessionId)); err != nil {
+		logger.Error().Err(err).Msg("failed to delete outdated session metadata")
+		return err
+	}
+
+	delete(rs.sessionsTrees[supplierOperatorAddress][sessionEndHeight], sessionId)
+
+	return nil
+}
+
+// persistSessionMetadata persists the session metadata to the store.
+// It is used to persist the session metadata using the sessionId as key,
+// after a session has been created.
+func (rs *relayerSessionsManager) persistSessionMetadata(sessionTree relayer.SessionTree) error {
+	sessionId := sessionTree.GetSessionHeader().SessionId
+	sessionEndHeight := sessionTree.GetSessionHeader().SessionEndBlockHeight
+	supplierOperatorAddress := sessionTree.GetSupplierOperatorAddress()
+
+	persistedSMT := &prooftypes.PersistedSMT{
+		SessionHeader:           sessionTree.GetSessionHeader(),
+		SupplierOperatorAddress: supplierOperatorAddress,
+		SmtRoot:                 sessionTree.GetSMSTRoot(),
+	}
+	logger := rs.logger.With(
+		"method", "persistSessionMetadata",
+		"supplier_operator_address", supplierOperatorAddress,
+		"session_id", sessionId,
+		"session_end_height", sessionEndHeight,
+	)
+
+	metadataBz, err := persistedSMT.Marshal()
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to marshal metadata")
+		return err
+	}
+
+	if err := rs.sessionsMetadataStore.Set([]byte(sessionId), metadataBz); err != nil {
+		logger.Error().Err(err).Msg("failed to persist session metadata")
+		return err
+	}
+
+	return nil
+}
+
+// insertSessionTree adds the given session to the sessionTrees map of
+// supplierOperatorAddress->blockHeight->sessionId->sessionTree
+func (rs *relayerSessionsManager) insertSessionTree(sessionTree relayer.SessionTree) bool {
+	sessionEndHeight := sessionTree.GetSessionHeader().SessionEndBlockHeight
+	sessionId := sessionTree.GetSessionHeader().SessionId
+	supplierOperatorAddress := sessionTree.GetSupplierOperatorAddress()
+
+	// Get the suppliersSessionTrees for the given supplier.
+	supplierSessionTrees, ok := rs.sessionsTrees[supplierOperatorAddress]
+
+	// If there is no map for session trees with the supplier operator address, create one.
+	if !ok {
+		supplierSessionTrees = make(map[int64]map[string]relayer.SessionTree)
+		rs.sessionsTrees[supplierOperatorAddress] = supplierSessionTrees
+	}
+
+	// Get the session trees for the given sessionEndHeight.
+	sessionTreesWithEndHeight, ok := supplierSessionTrees[sessionEndHeight]
+
+	// If there is no map for sessions at the sessionEndHeight, create one.
+	if !ok {
+		sessionTreesWithEndHeight = make(map[string]relayer.SessionTree)
+		supplierSessionTrees[sessionEndHeight] = sessionTreesWithEndHeight
+	}
+
+	// Get the sessionTree for the given session id.
+	if _, ok = sessionTreesWithEndHeight[sessionId]; ok {
+		return false
```

Comment:
WHy don't we allow overrides?

Note that I've updated the godoc comment, but still feel like this should be okay behaviour.

Reply:
Note that this map is `supplierOperatorAddress->blockHeight->sessionId->sessionTree`

When loading session trees from the store, there should be only one corresponding to a given supplier+sessionId.

Updating this case with:
```go
	// Get the sessionTree for the given session id.
	// An already existing session tree means that the the sessionTreeStore has
	// duplicate session trees for the same supplier operator address and session id,
	// which should not happen.
	if _, ok = sessionTreesWithEndHeight[sessionId]; ok {
		rs.logger.Warn().
			Str("session_id", sessionId).
			Str("supplier_operator_address", supplierOperatorAddress).
			Msg("session tree already exists, skipping")
		return false
	}
```